File: backend/app/services/pricing_service.py
import asyncpg
from datetime import datetime


async def get_active_tariff(conn: asyncpg.Connection) -> dict:
    row = await conn.fetchrow("SELECT * FROM tariffs WHERE is_active=true LIMIT 1")
    if not row:
        raise ValueError("No active tariff configured")
    return dict(row)


def calculate_price(distance_km: float, duration_min: float, tariff: dict) -> float:
    hour = datetime.now().hour
    is_night = hour >= 22 or hour < 6
    price = (
        float(tariff["base_fare"])
        + distance_km * float(tariff["per_km_price"])
        + duration_min * float(tariff["per_min_price"])
    )
    if is_night:
        price *= float(tariff["night_multiplier"])
    min_fare = float(tariff["min_fare"])
    rounded = round(price / 100) * 100
    return max(rounded, min_fare)


async def estimate_price(conn: asyncpg.Connection, distance_km: float, duration_min: float) -> tuple[float, dict]:
    tariff = await get_active_tariff(conn)
    return calculate_price(distance_km, duration_min, tariff), tariff

File: backend/src/services/pricing.service.js
const db = require('../config/database');

async function getActiveTariff() {
  const tariff = await db('tariffs').where({ is_active: true }).first();
  if (!tariff) throw new Error('No active tariff');
  return tariff;
}

function calculatePrice(distanceKm, durationMin, tariff) {
  const hour = new Date().getHours();
  const isNight = hour >= 22 || hour < 6;

  let price =
    parseFloat(tariff.base_fare) +
    distanceKm * parseFloat(tariff.per_km_price) +
    durationMin * parseFloat(tariff.per_min_price);

  if (isNight) price *= parseFloat(tariff.night_multiplier);

  return Math.max(Math.ceil(price / 100) * 100, parseFloat(tariff.min_fare));
}

async function estimatePrice(distanceKm, durationMin) {
  const tariff = await getActiveTariff();
  return { price: calculatePrice(distanceKm, durationMin, tariff), tariff };
}

module.exports = { estimatePrice, calculatePrice, getActiveTariff };
